rust/xaynet-analytics/src/database/isar.rs

```
@@ -0,0 +1,125 @@
+            });
+
+        // TODO: not sure how to proceed to parse [u8] using the collection schema. didn't find examples in Isar
+        unimplemented!()
+    }
+
+    pub fn put(&self, collection_name: &str, object: &[u8]) -> Result<(), Error> {
+        self.get_collection(collection_name)?
+            .put(&self.begin_txn(false)?, None, object)
+            .map_err(|_| {
+                anyhow!(
+                    "failed to add object {:?} to collection: {}",
+                    object,
+                    collection_name
+                )
+            })
+            .map(|_| ())
+    }
+
+    pub fn get_object_builder(&self, collection_name: &str) -> Result<ObjectBuilder, Error> {
+        Ok(self.get_collection(collection_name)?.get_object_builder())
+    }
+
+    fn get_schema() -> Result<Schema, Error> {
+        let mut schema = Schema::new();
+        schema
+            .add_collection(get_collection_schema(
+                Self::ANALYTICS_EVENT_NAME,
+                AnalyticsEvent::into_field_properties(),
+            )?)
+            .map_err(|_| {
+                anyhow!(
+                    "failed to add collection {} to schema",
+                    Self::ANALYTICS_EVENT_NAME
+                )
+            })
+            .map(|_| schema)
+    }
+
+    fn get_collection(&self, collection_name: &str) -> Result<&IsarCollection, Error> {
+        self.instance
+            .get_collection_by_name(collection_name)
+            .ok_or_else(|| anyhow!("wrong collection name: {}", collection_name))
+    }
+
+    fn begin_txn(&self, write: bool) -> Result<IsarTxn, Error> {
+        self.instance
+            .begin_txn(write)
+            .map_err(|_| anyhow!("failed to begin transaction"))
+    }
+}
+
+fn get_collection_schema(
+    name: &str,
+    field_properties: IntoIter<FieldProperty>,
+) -> Result<CollectionSchema, Error> {
+    let mut schema = CollectionSchema::new(&name);
+    field_properties.for_each(|prop| {
```

This is a `try_fold`, it will help in propagating the error. At the moment the error is just ignored you call `ok`.

---

yeh i know, i didn't manage to get `try_fold` to work. i'll try again..

---

ok done, thanks for the help on the call